androidenhancedvideoplayer/src/main/java/com/profusion/androidenhancedvideoplayer/components/playerOverlay/VolumeControl.kt

```diff
@@ -0,0 +1,88 @@
+package com.profusion.androidenhancedvideoplayer.components.playerOverlay
+
+import androidx.compose.foundation.interaction.MutableInteractionSource
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Spacer
+import androidx.compose.foundation.layout.fillMaxHeight
+import androidx.compose.foundation.layout.height
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import com.profusion.androidenhancedvideoplayer.styling.Dimensions
+
+private const val MIN_VOLUME_VALUE = 0
+
+enum class VolumeIconState {
+    HIGH,
+    LOW,
+    OFF
+}
+
+fun mapVolumeRangeToState(volume: Int, maxVolumeValue: Int): VolumeIconState {
+    val halfVolume = maxVolumeValue / 2
+    val lowRange = MIN_VOLUME_VALUE + 1..halfVolume
+    val highRange = halfVolume + 1..maxVolumeValue
+    return when (volume) {
+        in lowRange -> VolumeIconState.LOW
+        in highRange -> VolumeIconState.HIGH
+        else -> VolumeIconState.OFF
+    }
+}
+
+fun mapVolumeIconStateToIcon(
+    state: VolumeIconState,
+    customization: ControlsCustomization
+): @Composable () -> Unit {
+    return when (state) {
+        VolumeIconState.LOW -> customization.volumeLowIconContent
+        VolumeIconState.OFF -> customization.volumeOffIconContent
+        VolumeIconState.HIGH -> customization.volumeHighIconContent
+    }
+}
+
+@Composable
+fun VolumeControl(
+    modifier: Modifier = Modifier,
+    shouldEnableVolumeControl: Boolean,
+    isFullScreen: Boolean,
+    volume: () -> Int,
+    maxVolumeValue: () -> Int,
+    setDeviceVolume: (Int) -> Unit,
+    customization: ControlsCustomization,
+    volumeMutableInteractionSource: MutableInteractionSource
+) {
+    if (isFullScreen && shouldEnableVolumeControl) {
+        var volumeIconState by remember {
+            mutableStateOf(
+                mapVolumeRangeToState(volume = volume(), maxVolumeValue = maxVolumeValue())
+            )
+        }
+        val volumeIcon = mapVolumeIconStateToIcon(volumeIconState, customization)
+
+        Column(
+            modifier = modifier.fillMaxHeight(),
+            verticalArrangement = Arrangement.Center,
+            horizontalAlignment = Alignment.CenterHorizontally
+        ) {
+            volumeIcon()
+            Spacer(modifier = Modifier.height(Dimensions.xxlarge))
+            VerticalSlider(
+                initialValue = volume().toFloat(),
+                valueRange = MIN_VOLUME_VALUE.toFloat()..maxVolumeValue().toFloat(),
+                onValueChange = {
+                    setDeviceVolume(it.toInt())
+                    volumeIconState = mapVolumeRangeToState(
```

Comment:
Don't update the volumeIconState here. Update it based on `volume()` as the unique source of true.